src/lib/io/sol-uart-impl-linux.c

```diff
@@ -87,10 +86,10 @@ read_timeout(void *data)
     bool keep_running = true;
     int err;
 
-    uart->in_use = true;
-    r = uart->on_data((void *)uart->user_data, uart, &uart->rx);
-    uart->in_use = false;
-    SOL_INT_CHECK_GOTO(r, < 0, exit);
+    SOL_REENTRANT_CALL(uart->reentrant) {
+        r = uart->on_data((void *)uart->user_data, uart, &uart->rx);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
```

Comment:
You can't jump out of the body of a ```SOL_REENTRANT_CALL()```, because it has an epilogue which must not be skipped, or you'll leak the structure it's protecting.